o2a/transformers/add_workflow_notificaton_transformer.py

```diff
@@ -0,0 +1,86 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Remove Start Transformer
+"""
+from o2a.converter.parsed_action_node import ParsedActionNode
+from o2a.converter.task import Task
+from o2a.converter.workflow import Workflow
+from o2a.o2a_libs.property_utils import PropertySet
+from o2a.transformers.base_transformer import BaseWorkflowTransformer
+
+PROP_WORKFLOW_NOTIFICATION_URL = "oozie.wf.workflow.notification.url"
+
+
+def _get_nodes_without_upstream(workflow):
+    nodes = []
+    for node in workflow.nodes.values():
+        upstream_nodes = workflow.find_upstream_nodes(node)
+        if not upstream_nodes:
+            nodes.append(node)
+    return nodes
+
+
+def _get_nodes_without_downstream(workflow):
+    nodes = []
+    for node in workflow.nodes.values():
+        if not node.downstream_names and not node.error_xml:
+            nodes.append(node)
+    return nodes
+
+
+# pylint: disable=too-few-public-methods
+class AddWorkflowNotificationTransformer(BaseWorkflowTransformer):
+    """
+    Add workflow notification nodes
+    """
+
+    def process_workflow(self, workflow: Workflow, props: PropertySet):
+        notification_url = props.job_properties.get(PROP_WORKFLOW_NOTIFICATION_URL, None)
+        if not notification_url:
+            return
+
+        nodes_without_upstream = _get_nodes_without_upstream(workflow)
+        nodes_without_downstream = _get_nodes_without_upstream(workflow)
+
+        start_workflow_node = self.create_notification_node(notification_url, "start_workflow", "RUNNING")
+        end_success_workflow_node = self.create_notification_node(
+            notification_url, "end_success_workflow", "SUCCEEDED"
+        )
+        end_fail_workflow_node = self.create_notification_node(
+            notification_url, "end_fail_workflow", "FAILED"
+        )
+
+        workflow.nodes[start_workflow_node.name] = start_workflow_node
+        workflow.nodes[end_success_workflow_node.name] = end_success_workflow_node
+        workflow.nodes[end_fail_workflow_node.name] = end_fail_workflow_node
+
+        start_workflow_node.downstream_names = [node.name for node in nodes_without_upstream]
+        for node in nodes_without_downstream:
+            if not node.downstream_names:
+                node.downstream_names.append(end_success_workflow_node.name)
+            if not node.error_xml:
+                node.error_xml = end_fail_workflow_node.name
+
+    @staticmethod
+    def create_notification_node(url_template, start_task_name, status):
+        url = (
+            url_template.replace("$parentId", "").replace("$jobId", "{{ run_id }}").replace("$status", status)
+        )
+        start_workflow_node = ParsedActionNode(
```

Comment:
As tempting as it is, I don't quite like creating the start notification node of the `ParsedActionNode`.
This is rather misleading, as this is not in fact a "parsed node", coming from the XML, but a created one.
Maybe we should think about some class structure of `Node`s to represent the situation more clearly. Something akin to:
- Node
  -- ParsedActionNode
  -- NotificationNode
(just a suggestion)

What do you think? I believe it would be clearer.

Reply:
We discussed this topic offline. We chose a different solution.